src/hotspot/share/runtime/serviceThread.cpp

```diff
@@ -59,16 +59,25 @@ JvmtiDeferredEvent* ServiceThread::_jvmti_event = NULL;
 // to add this field to the per-JavaThread event queue.  TODO: fix this sometime later
 JvmtiDeferredEventQueue ServiceThread::_jvmti_service_queue;
 
-// Defer releasing JavaThread OopHandle to the ServiceThread
+// Defer releasing JavaThread OopHandles to the ServiceThread
 class OopHandleList : public CHeapObj<mtInternal> {
-  OopHandle      _handle;
+  static const int _count = 4;  // Need to keep in sync with JavaThread
+  OopHandle _handles[_count];
   OopHandleList* _next;
+  int _index;
  public:
-   OopHandleList(OopHandle h, OopHandleList* next) : _handle(h), _next(next) {}
-   ~OopHandleList() {
-     _handle.release(JavaThread::thread_oop_storage());
-   }
-   OopHandleList* next() const { return _next; }
+  OopHandleList(OopHandleList* next) : _next(next), _index(0) {}
+  void add(OopHandle h) {
+    assert(_index < _count, "too many additions");
```

Comment:
I think this should be a guarantee.

Reply:
Why? This is internally used code and any usage bugs should be exposed by testing. There is very little that can go wrong here.

Reply:
First to signal to a reader that the VM is utterly broken if this happens.
Secondly when running test we also run on release, having the guarantee makes triage this failure to same as the assert much faster.

Keep assert if you want.

Reply:
I prefer to keep the assert. I think guarantees should be reserved for things that are unexpected/highly-unlikely but still remotely possible due to external code actions - which is not the case here. Else every assert becomes a guarantee.